### PR TITLE
Robodummy

### PR DESCRIPTION
**Pull request by Pranjal-sopho**

### CODE REVIEW

File: mr_robot/.gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+*nn
+*hdf
```
Comment (FynnBe) [inline mr_robot/.gitignore]: There is no need for a second .gitignore file

File: mr_robot/mr_robot.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    def dense_annotate(self, x, y, label, image):
+        raise NotImplementedError
+
+    def terminate():
+        new_server.shutdown()
+
+class BaseStrategy:
+
+    def __init__():
+        raise NotImplementedError
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: `__init__` should not raise `NotImplementedError`. In fact, it is a good style to call `super().__init__()` in your derived class...

Comment (FynnBe) [reply mr_robot/mr_robot.py]: actually, this being work in progress, if you want to indicate that your `BaseStrategy` is not fully implemented yet, this is fine. (Calling `super().__init__()` in your derived class would still make sense)

File: mr_robot/utils.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    print("Non-trainable params: {0:,}".format(total_params - trainable_params))
+    print("----------------------------------------------------------------")
+    print("Input size (MB): %0.2f" % total_input_size)
+    print("Forward/backward pass size (MB): %0.2f" % total_output_size)
+    print("Params size (MB): %0.2f" % total_params_size)
+    print("Estimated Total Size (MB): %0.2f" % total_size)
+    print("----------------------------------------------------------------")
+
+
+def make_edges3d(segmentation):
```
Comment (FynnBe) [inline mr_robot/utils.py]: As this function comes from https://github.com/constantinpape/vis_tools/blob/master/vis_tools/edges.py#L5 I would add a reference here (as comment)

File: mr_robot/utils.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    """
+    # NOTE we add one here to make sure that we don't have zero in the segmentation
+    gz = convolve(segmentation + 1, np.array([-1.0, 0.0, 1.0]).reshape(3, 1, 1))
+    gy = convolve(segmentation + 1, np.array([-1.0, 0.0, 1.0]).reshape(1, 3, 1))
+    gx = convolve(segmentation + 1, np.array([-1.0, 0.0, 1.0]).reshape(1, 1, 3))
+    return (gx ** 2 + gy ** 2 + gz ** 2) > 0
+
+# create patches
+def tile_image2D(image_shape, tile_size):
+
```
Comment (FynnBe) [inline mr_robot/utils.py]: it seems to me that image tiling could nicely be implemented for n dimensions. Maybe have a look at https://github.com/ilastik/lazyflow/blob/dfbb450989d4f790f5b19170383b777fb88be0e8/lazyflow/roi.py#L473 for some inspiration

File: mr_robot/mr_robot.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
+# import sys
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: we should avoid uncommented import statements (just remove this line)

File: mr_robot/mr_robot.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        with open("state.nn", mode="rb") as f:
+            binary_state = f.read()
+        with open("model.py", mode="rb") as f:
+            model_file = f.read()
+
+        with open("robo_config.yml", mode="r") as f:
+            base_config = yaml.load(f)
+
+        fut = self.new_server.load_model(base_config, model_file, binary_state, b"", ["cpu"])
+        print("model loaded")
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: [optional] use a logger, instead of print:
```
import logging
logger = logging.getLogger(__name__)
...
logger.info("model loaded")
```
more at https://docs.python.org/3/howto/logging.html

File: mr_robot/mr_robot.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        # print(fut.result())
+
+    def resume(self):
+        self.new_server.resume_training()
+        print("training resumed")
+
+    def predict(self):
+        self.ip = np.expand_dims(self.f['volume'][0,0:img_dim, 0:img_dim], axis = 0)
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: instead of taking the first slice [0, ...] and then expanding the resulting array, you should simplify to take a slice right away:
[0:1, ...]

File: mr_robot/mr_robot.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+if __name__ == "__main__":
+    
+    robo = MrRobot()
+    file = robo.load_data()
+    robo.load_model()
+    robo.resume()  # resume training
+
+    # run prediction
+    op = robo.predict()
+
```
Comment (m-novikov) [inline mr_robot/mr_robot.py]: Here I think algorithm should be read as follows:
```python
# Step 1. Intialization
robo = MrRobot('/home/user/config.yaml')  # Here robot loads all required data
robo.use_strategy(StrategyRandom())
# or even
robo = MrRobot('/home/user/config.yaml', StrategyRandom)

# Step 2. Start
robo.start()  # Start tiktorch server

# Step 3. Prediction loop
while robo.should_stop():
      robo.predict()

# def robo.predict
# 1. labels? = self.strategy.get_next_patch(<relevant data>)
# 2. self.update_training(labels, ...)

# Step 4. Termination
robo.terminate()

File: mr_robot/mr_robot.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+class Strategy3(BaseStrategy):
+    def __init__():
+        super().__init__()
+        raise NotImplementedError
+
+
+if __name__ == "__main__":
+
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: the following code should be inside of MrRobot. Currently you mirror parts of the tiktorch api in MrRobot (methods: resume, predict, add). This is fine for convenience, etc, but in it's core MrRobot should implement the way of running a 'user simulation'

File: mr_robot/mr_robot.py [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+class BaseStrategy:
+    def __init__(self, file, op):
+        self.f = file
+        self.op = op
+
+    # compute loss for a given patch
+    def base_loss(self, patch, label):
+        label = label[0][0]
+        patch = patch[0][0]
+        result = mean_squared_error(label, patch)  # CHECK THIS
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: the criterion should be configurable

File: mr_robot/mr_robot.py [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+        return result
+
+
+class Strategy1(BaseStrategy):
+    def __init__(self, file, op):
+        super().__init__(file,op)
+
+    def run(self):
+        idx = tile_image(self.op.shape, patch_size)
+        label = np.expand_dims(self.f['volumes/labels/neuron_ids'][0,0:img_dim,0:img_dim], axis=0)
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: same indexing as in predict method

File: mr_robot/mr_robot.py [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+        patch = patch[0][0]
+        result = mean_squared_error(label, patch)  # CHECK THIS
+        return result
+
+
+class Strategy1(BaseStrategy):
+    def __init__(self, file, op):
+        super().__init__(file,op)
+
+    def run(self):
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: I would prefer the robot class to perform the 'run', not the strategy. The strategy should effectively implement a sampling strategy. I see this analog to the [pytorch sampler](https://pytorch.org/docs/stable/_modules/torch/utils/data/sampler.html).
We might even be able to use the [pytorch dataset](https://pytorch.org/docs/stable/data.html#torch.utils.data.Dataset) and the [pytorch dataloader](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader) for our purposes (and then implement our strategy as a 'Sampler'

File: mr_robot/mr_robot.py (hunk 8 above)
Comment (FynnBe) [reply mr_robot/mr_robot.py]: yes, I'd vote for
> `robo = MrRobot('/home/user/config.yaml', StrategyRandom)`

File: mr_robot/mr_robot.py [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+class MrRobot:
+    def __init__(self, path_to_config_file, strategy):
+        # start the server
+        self.new_server = TikTorchServer()
+        self.strategy =strategy
+
+        with open(path_to_config_file, mode="r") as f:
+            self.base_config = yaml.load(f)
+
+        self.max_robo_iterations = self.base_config['max_robo_iterations']
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: including `_robo_` in this variable name seems redundant, considering we are in the `MrRobot` class

File: mr_robot/mr_robot.py [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+        self.counter = 0
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: `counter` as a property name is a bit confusing here (it is not obvious what's being counted)
I would suggest
- `self.iterations_max`
- `self.iterations_done`

if that is what you intend

File: mr_robot/mr_robot.py [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+
+
+    def load_model(self):
+        # load the model
+
+        
+        #with open(base_config['cremi_data_dir'], mode="rb") as f:
+        #    binary_state = f.read()
+
+        archive = zipfile.ZipFile(self.base_config['cremi_dir']['path_to_zip'], 'r')
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: you should run black on your code (this will convert `'` to `"` where possible)

File: mr_robot/mr_robot.py [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+            return True
+
+    def run(self):
+        self.strategy.patch('MSE', self.op)
+        while(self.stop()):
+            idx = self.strategy.get_next_patch(self.op)
+            self.add(idx)
+
+    def add(self, idx):
+        file = z5py.File(self.base_config["cremi_data"])
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: no need to open this file every time `add` is called. This should be in `__init__`. use `self.file` here

Comment (FynnBe) [inline mr_robot/mr_robot.py]: `"cremi_data"` should instead be something like `"raw_data_path"` and `"label_data_path"` (1. cremi is just an example. 2. raw data and label data are not necessarily in the same file)

File: mr_robot/mr_robot.py [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+        labels = file["cremi_path_to_labelled"][0:1, 0:img_dim, 0:img_dim]
+
+        new_ip = self.ip.as_numpy()[idx[0]:idx[1], idx[2]:idx[3], idx[4]:idx[5]].astype(float)
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: you should not hardcode that the data is 3 dimensional, use tuples to index instead

File: mr_robot/mr_robot.py [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+        self.logger = logging.getLogger(__name__)
+   
+    def loss(self,tile,label, loss_fn):
+        label = label[0]
+        tile = tile[0]
+        result = mean_squared_error(label, tile)  # CHECK THIS
+        return result
+
+    def base_patch(self, loss_fn, op):
+        idx = tile_image(op.shape, patch_size)
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: it would be great if you could add some doc strings to communicate what your methods (and classes) are for

Comment (FynnBe) [reply mr_robot/mr_robot.py]: hint: https://www.python.org/dev/peps/pep-0008/#documentation-strings

Comment (FynnBe) [reply mr_robot/mr_robot.py]: also: no need to call this method `base_patch`. naming it `patch` and calling `super().patch()` in a derived class works as well (even when you overwrite `patch` in the derived class, that's what the `super()`, resolves for you) 
some more 'how-to-inherit' here: https://www.python.org/dev/peps/pep-0008/#designing-for-inheritance

File: lala.py [hunk 19]
```diff
@@ -0,0 +1,6 @@
+import yaml
```
Comment (m-novikov) [inline lala.py]: Please delete this file

Comment (m-novikov) [reply lala.py]: Also `neuronid.txt` seems unused so it is also a candidate for deletion from this pull request.

**Comment by m-novikov** [comment]
You should create `__init__.py` within `mr_robot` directory to make it proper python package. Now it is implicit namespace which is inconvenient in 9/10 cases.

File: mr_robot/utils.py [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+            summary[m_key]["input_shape"][0] = batch_size
+            if isinstance(output, (list, tuple)):
+                summary[m_key]["output_shape"] = [[-1] + list(o.size())[1:] for o in output]
+            else:
+                summary[m_key]["output_shape"] = list(output.size())
+                summary[m_key]["output_shape"][0] = batch_size
+
+            params = 0
+            if hasattr(module, "weight") and hasattr(module.weight, "size"):
+                params += torch.prod(torch.LongTensor(list(module.weight.size())))
```
Comment (m-novikov) [inline mr_robot/utils.py]: This function seems strange, it uses `torch` but `torch` is not imported.

File: mr_robot/mr_robot.py [hunk 21]
```diff
@@ -0,0 +1,214 @@
...
+import yaml
+import logging
+from abc import ABC, abstractmethod
+from tensorboardX import SummaryWriter
+from tiktorch.server import TikTorchServer
+from tiktorch.rpc import Client, Server, InprocConnConf
+from tiktorch.rpc_interface import INeuralNetworkAPI
+from tiktorch.types import NDArray, NDArrayBatch
+from tests.conftest import nn_sample
+from mr_robot.utils import *
```
Comment (m-novikov) [inline mr_robot/mr_robot.py]: I would recommend avoiding `*` imports it's hard to follow. Some functions in utils are broken and because of `*` import, one can't see if they are used somewhere in this file.

File: mr_robot/mr_robot.py [hunk 22]
```diff
@@ -0,0 +1,214 @@
...
+import zipfile
+import h5py
+import z5py
+from z5py.converter import convert_from_h5
+from scipy.ndimage import convolve
+from torch.autograd import Variable
+from collections import OrderedDict
+import yaml
+import logging
+from abc import ABC, abstractmethod
```
Comment (m-novikov) [inline mr_robot/mr_robot.py]: This stuff should be introduced at later stages of the project if at all. It infects code and spreads like a plague, without improving developer experience.

File: mr_robot/mr_robot.py [hunk 23]
```diff
@@ -0,0 +1,214 @@
...
+from tensorboardX import SummaryWriter
+from tiktorch.server import TikTorchServer
+from tiktorch.rpc import Client, Server, InprocConnConf
+from tiktorch.rpc_interface import INeuralNetworkAPI
+from tiktorch.types import NDArray, NDArrayBatch
+from tests.conftest import nn_sample
```
Comment (m-novikov) [inline mr_robot/mr_robot.py]: Should not be there

File: mr_robot/mr_robot.py [hunk 24]
```diff
@@ -0,0 +1,214 @@
...
+
+        criterion_class = getattr(nn, loss_fn, None)
+        assert criterion_class is not None, "Criterion {} not found.".format(method)
+
+        curr_loss = criterion_class(torch.from_numpy(tile), torch.from_numpy(target))
+        self.patched_data.append((curr_loss, slicer))
+
+    @abstractmethod
+    def get_next_patch(self):
+        pass
```
Comment (m-novikov) [inline mr_robot/mr_robot.py]: Use `raise NotImplementedError()` instead of abstract method

File: tiktorch/handler/script.py [hunk 25]
```diff
@@ -0,0 +1,10 @@
+import sys
```
Comment (m-novikov) [inline tiktorch/handler/script.py]: Maybe this file is also candidate for deletion

File: mr_robot/utils.py [hunk 26]
```diff
@@ -0,0 +1,140 @@
...
+    """
+    # NOTE we add one here to make sure that we don't have zero in the segmentation
+    gz = convolve(segmentation + 1, np.array([-1.0, 0.0, 1.0]).reshape(3, 1, 1))
+    gy = convolve(segmentation + 1, np.array([-1.0, 0.0, 1.0]).reshape(1, 3, 1))
+    gx = convolve(segmentation + 1, np.array([-1.0, 0.0, 1.0]).reshape(1, 1, 3))
+    return (gx ** 2 + gy ** 2 + gz ** 2) > 0
+
+
+# create patches
+def tile_image(image_shape, tile_size):
```
Comment (m-novikov) [inline mr_robot/utils.py]: This function seems like a good candidate for proper tests.

File: mr_robot/mr_robot.py [hunk 27]
```diff
@@ -0,0 +1,214 @@
...
+        # start the server
+        self.new_server = TikTorchServer()
+        self.strategy = strategy
+
+        with open(path_to_config_file, mode="r") as f:
+            self.base_config = yaml.load(f)
+
+        self.data_file = z5py.File(self.base_config.pop("raw_data_path"))
+
+        self.indices = tile_image(self.base_config["training"]["training_shape"], patch_size)
```
Comment (m-novikov) [inline mr_robot/mr_robot.py]: It looks like `tile_image` should not only return starting indices for a patch but also last ones(maybe even slices?), usage of `self.indicies` requires a user to know about patch_size
Also `self.indicies` is undescriptive name, should be at least something like `self.tile_indicies`

File: mr_robot/mr_robot.py [hunk 28]
```diff
@@ -0,0 +1,214 @@
...
+
+        self.patched_data = []
+        self.logger = logging.getLogger(__name__)
+
+    def _loss(self, op, target, loss_fn, slicer):
+        """  computes loss corresponding to the output and target according to 
+        the given loss function
+
+        Args:
+        op(np.ndarray) : predicted output
```
Comment (m-novikov) [inline mr_robot/mr_robot.py]: non-descriptive names are bad

File: mr_robot/mr_robot.py [hunk 29]
```diff
@@ -0,0 +1,214 @@
...
+        target(np.ndarray): ground truth
+        loss_fn(string): loss metric
```
Comment (m-novikov) [inline mr_robot/mr_robot.py]: IMO `loss_fn` should be a part of the strategy, @FynnBe what do you think?

File: mr_robot/mr_robot.py [hunk 30]
```diff
@@ -0,0 +1,214 @@
...
+            self.strategy.rearrange()
+            slicer = self.strategy.get_next_patch(self.op)
+            self.indices.pop(self.patch_id[slicer])
+            self._add(slicer)
+            self._resume()
+
+        self.terminate()
+
+    def _add(self, slicer):
+        new_ip = self.ip.as_numpy()[slicer].astype(float)
```
Comment (m-novikov) [inline mr_robot/mr_robot.py]: bad name (i know only two possible interpretations: Internet Protocol, Intelectual Property) and I don't see where `self.ip` is even set

File: tests/robot/test_mr_robot.py [hunk 31]
```diff
@@ -0,0 +1,24 @@
+import numpy as np
+from mr_robot.mr_robot import MrRobot
+from tiktorch.server import TikTorchServer
+from tiktorch.rpc import RPCFuture
+from tiktorch.types import SetDeviceReturnType
+
+import z5py
+
+def test_MrRobot():
```
Comment (m-novikov) [inline tests/robot/test_mr_robot.py]: this test won't run

**Comment by m-novikov** [comment]
Seems like it starting to take shape.
We should ensure that we have at least a couple of running tests; the current one seems like it's not working. Ping me if you want some help with this. We can do pair programming or something.

File: tests/robot/test_mr_robot.py (hunk 31 above)
Comment (Pranjal-sopho) [reply tests/robot/test_mr_robot.py]: why not? i have been using it to test the whole afternoon....

Comment (m-novikov) [reply tests/robot/test_mr_robot.py]: maybe some commits are not pushed to your repo then:
```
    def test_MrRobot():
>       robo = MrRobot()
E    TypeError: __init__() missing 2 required positional arguments: 'path_to_config_file' and 'strategy'

tests/robot/test_mr_robot.py:10: TypeError
```
this is what i get

File: mr_robot/mr_robot.py (hunk 27 above)
Comment (Pranjal-sopho) [reply mr_robot/mr_robot.py]: 1. transfer the patch_size variable to the config file
2. change variable name to self.tile_start_indices.
does this sound better?

File: mr_robot/mr_robot.py (hunk 22 above)
Comment (Pranjal-sopho) [reply mr_robot/mr_robot.py]: sorry, can you please elaborate a bit on this?

Comment (m-novikov) [reply mr_robot/mr_robot.py]: What are the benefits of abs usage compared to `raise NotImplementedError()` for this project?
Usage of `abc` also requires introduction ABC class to the inheritance hierarchy and we don't actually have a stable API usage of which we want to enforce. So `abc` usage looks unnecessary.

File: mr_robot/mr_robot.py (hunk 27 above)
Comment (m-novikov) [reply mr_robot/mr_robot.py]: Can we have a tile with attributes start and end? Why do we users of this function should know about patch size?

Comment (FynnBe) [reply mr_robot/mr_robot.py]: actually `tiktorch_config["training"]["training_shape"]` is the `patch_size`.
The robot reads this out and knows about it, if it is provided by the robot where needed, I'd totally be fine with only using the staring indices as tile IDs.
To compute the tiles the training image shape needs to be determined. This should be done by inspecting whatever training data is specified in the `robot` seciont of `tiktorch_config`

File: mr_robot/mr_robot.py [hunk 32]
```diff
@@ -0,0 +1,214 @@
...
+        raise NotImplementedError
+
+    def terminate(self):
+        self.tensorboard_writer.close()
+        self.new_server.shutdown()
+
+
+class BaseStrategy(ABC):
+    def __init__(self, path_to_config_file):
+        with open(path_to_config_file, mode="r") as f:
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: `MrRobot` should read the config file and initialize a strategy with the relevant parameters (or at least the config as a dict)

File: mr_robot/mr_robot.py (hunk 29 above)
Comment (FynnBe) [reply mr_robot/mr_robot.py]: warning: I went back and forth on this one already....
I think we can abstract way the loss_fn from the strategy. It should, however, be set in `__init__` (and optimally 'declared' in the class for type annotations). We can take it's value from the `robot` section in `tiktorch_config` and maybe use the training loss as a default

File: mr_robot/mr_robot.py [hunk 33]
```diff
@@ -0,0 +1,218 @@
...
+from collections import OrderedDict
+import yaml
+import logging
+from tensorboardX import SummaryWriter
+from tiktorch.server import TikTorchServer
+from tiktorch.rpc import Client, Server, InprocConnConf
+from tiktorch.rpc_interface import INeuralNetworkAPI
+from tiktorch.types import NDArray, NDArrayBatch
+from mr_robot.utils import tile_image
+
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: maybe sort the import statements a little don't mix import... and from... too much

File: mr_robot/mr_robot.py [hunk 34]
```diff
@@ -0,0 +1,218 @@
...
+        self.tile_indices = tile_image(self.base_config["training"]["training_shape"], patch_size)
+        self.input_shape = list((self.base_config["training"]["training_shape"]))
+        self.slicer = [slice(0, i) for i in self.input_shape]
+
+        self.iterations_max = self.base_config.pop("max_robo_iterations")
+        self.iterations_done = 0
+        self.tensorboard_writer = SummaryWriter()
+        self.logger = logging.getLogger(__name__)
+        plt.ion()
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: remove this?

File: mr_robot/mr_robot.py [hunk 35]
```diff
@@ -0,0 +1,218 @@
...
+
+    # def load_data(self):
+    #    self.f = z5py.File(self.base_config["cremi_data_dir"])
+    #    self.logger("data file loaded")
+
+    def _load_model(self):
+
+        archive = zipfile.ZipFile(self.base_config["data_dir"]["path_to_zip"], "r")
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: add "path_to_folder" and ability to load model and binary_state from (unzipped) folder as well.

Comment (FynnBe) [reply mr_robot/mr_robot.py]: or yet, check if path_to_folder ends in ".zip" or alike... and then use zipfile...

File: mr_robot/mr_robot.py [hunk 36]
```diff
@@ -0,0 +1,218 @@
...
+            self.iterations_done += 1
+            return True
+
+    def _run(self):
+        """ Feed patches to tiktorch (add to the training data)
+
+        The function fetches the patches in order decided by the strategy, 
+        removes it from the list of indices and feeds it to tiktorch 
+        """
+        while self.stop():
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: while not self.stop() ?

File: mr_robot/mr_robot.py [hunk 37]
```diff
@@ -0,0 +1,218 @@
...
+            self._predict()
+
+            # log average loss for all patches per iteration to tensorboard
+            total_loss = sum([pair[0] for pair in self.strategy.patched_data])
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: suggestion for better readability: 
`total_loss = sum([loss for loss, slice_ in self.strategy.patched_data])`

File: mr_robot/mr_robot.py [hunk 38]
```diff
@@ -0,0 +1,218 @@
...
+            avg = total_loss / float(len(self.strategy.patched_data))
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: alternatively use: `numpy.mean`

File: mr_robot/mr_robot.py [hunk 39]
```diff
@@ -0,0 +1,218 @@
...
+
+        for i in self.tile_indices:
+            self.slicer[-1] = slice(i[0][1], i[1][1])
+            self.slicer[-2] = slice(i[0][0], i[1][0])
+            new_slicer = tuple(self.slicer)
+            self.patch_id[i[0]] = x  # map each slicer with its corresponding index
+            x += 1
+            pred_output = self.new_server.forward(NDArray(self.data_file["volume"][new_slicer]))
+            pred_output = pred_output.result()
+            self.strategy._loss(
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: avoid calling 'private' methods externally

Comment (FynnBe) [reply mr_robot/mr_robot.py]: why does 'predict' need to call strategy.loss()?

File: mr_robot/mr_robot.py (hunk 33 above)
Comment (m-novikov) [reply mr_robot/mr_robot.py]: maybe https://github.com/thijsdezoete/sublime-text-isort-plugin/blob/master/README.md

File: tests/robot/test_mr_robot.py [hunk 40]
```diff
@@ -0,0 +1,32 @@
+import pytest
+from mr_robot.mr_robot import MrRobot, Strategy1
+from mr_robot.utils import tile_image
+from tiktorch.server import TikTorchServer
+
+
+def test_tile_image():
+    # when image dim are not multiple of patch size
+    tiled_indices = tile_image((1, 48, 48), 32)
+    assert len(tiled_indices) == 4
```
Comment (m-novikov) [inline tests/robot/test_mr_robot.py]: would be nice to also have the comparison with the exact result, not only length

File: mr_robot/mr_robot.py [hunk 41]
```diff
@@ -105,39 +112,36 @@ def stop(self):
...
         """
-        while self.stop():
+        while not self.stop():
             self._predict()
 
             # log average loss for all patches per iteration to tensorboard
-            total_loss = sum([pair[0] for pair in self.strategy.patched_data])
-            avg = total_loss / float(len(self.strategy.patched_data))
+            avg = np.mean([loss for loss, slice_ in self.strategy.patched_data])
             self.tensorboard_writer.add_scalar("avg_loss", avg, self.iterations_done)
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: to entagle MrRobot and a strategy futher, let the strategy compute the current loss mean and access it from withing MrRobot

File: mr_robot/mr_robot.py [hunk 42]
```diff
@@ -105,39 +112,36 @@ def stop(self):
...
 
-            self.strategy.rearrange()
-            slicer = self.strategy.get_next_patch()
-            self.tile_indices.pop(self.patch_id[(slicer[-2].start, slicer[-1].start)])
-            self._add(slicer)
+            block_batch = self.strategy.get_next_batch(batch_size)
+            for block in block_batch:
+                self.block_list.pop(self.patch_id[block[0].start])
+                self._add(block)
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: avoid dissembling and reassembling batches (update the batched data)

File: tests/robot/test_mr_robot.py [hunk 43]
```diff
@@ -0,0 +1,69 @@
...
+    ]
+
+    # when image too small for the patch
+    with pytest.raises(AssertionError):
+        tiled_indices = tile_image((1, 48, 48), (1, 64, 32))
+        tiled_indices = tile_image((64, 64), (2, 1, 1))
+
+
+def test_MrRobot():
+    os.environ["CUDA_VISIBLE_DEVICES"] = "7"
```
Comment (FynnBe) [inline tests/robot/test_mr_robot.py]: setting environment variables should not be hardcoded.
Set them in your terminal 
on windows:
```
set CUDA_VISIBLE_DEVICES=7
```
on linux
```
export CUDA_VISIBLE_DEVICES=7
```

File: .gitignore [hunk 44]
```diff
@@ -7,4 +7,6 @@ tiktorch/.idea
 tiktorch/__pycache/
 /#wrapper.py#
 /.#wrapper.py#
-.py~
\ No newline at end of file
+.py~
+*.nn
+*.hdf
```
Comment (FynnBe) [inline .gitignore]: there is no need to ignore `.nn` and `.hdf` files (as there are none in the repo). Pls remove

File: mr_robot/mr_robot.py [hunk 45]
```diff
@@ -0,0 +1,601 @@
...
+        coordinate.append(slice_.start)
+
+    return tuple(coordinate)
+
+
+class MrRobot:
+    """ The robot class runs predictins on the model, and feeds the
+    worst performing patch back for training. The order in which patches
+    are feed back is determined by the 'strategy'. The robot can change
+    strategies as training progresses.
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: we decided that changing strategy is a strategy of its own...

File: mr_robot/mr_robot.py [hunk 46]
```diff
@@ -0,0 +1,601 @@
...
+        print()
+
+        strategy_class = strategies[strategy]
+        self.strategy = strategy_class(
+            self.base_config["training"]["loss_criterion_config"]["method"], self.base_config["class_dict"]
+        )
+
+        self.iterations_max = self.base_config.pop("max_robo_iterations")
+        self.iterations_done = 0
+        self.tensorboard_writer = SummaryWriter(logdir="/home/psharma/psharma/repos/tiktorch/tests/robot/robo_logs")
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: do not hard code 'personal' paths, etc...
suggestion:
get absolute path of `mr_robot.py` and deduct the absolute path to `mr_robot` folder:
```
mr_robot_folder = os.path.dirname(os.path.abspath(__file__))
```
add log folder to it:
```
logdir=os.path.join(mr_robot_folder, "logs")
```

File: mr_robot/mr_robot.py (hunk 29 above)
Comment (FynnBe) [reply mr_robot/mr_robot.py]: please update doc string...

File: mr_robot/utils.py [hunk 47]
```diff
@@ -0,0 +1,154 @@
...
+            j += block_shape[n]
+
+    for i in range(no_of_blocks):
+        for j in range(len(block_list[i])):
+            block_list[i][j] = slice(block_list[i][j][0], block_list[i][j][1])
+
+        block_list[i] = tuple(block_list[i])
+
+    return block_list
+"""
```
Comment (FynnBe) [inline mr_robot/utils.py]: delete if this is no longer needed

File: mr_robot/mr_robot.py [hunk 48]
```diff
@@ -0,0 +1,601 @@
...
+                os.path.join(
+                    self.base_config["model_dir"]["path_to_folder"],
+                    self.base_config["model_dir"]["path_in_folder_to_state"],
+                )
+            )
+
+        # cleaning dictionary before passing to tiktorch
+        self.base_config.pop("model_dir")
+
+        self.new_server.load_model(self.base_config, model, binary_state, b"", ["gpu:4"])
```
Comment (FynnBe) [inline mr_robot/mr_robot.py]: do not hard code use of a specific gpu, use environment variables. 
you assert in  `__init__` that there is only one gpu available, thus you could hard code gpu:0 here. (the indices are mapped, e.g. if `export CUDA_VISIBLE_DEVICES=4,6` then available torch.cuda indices are: `[0, 1]`
